test/builder/pipeline.go

```diff
@@ -0,0 +1,185 @@
+// Any number of Pipeline modifier can be passed to transform it.
+func Pipeline(name string, ops ...PipelineOp) *v1alpha1.Pipeline {
+	p := &v1alpha1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: defaultNamespace,
+			Name:      name,
+		},
+	}
+
+	for _, op := range ops {
+		op(p)
+	}
+
+	return p
+}
+
+// PipelineNamespace sets the namespace of the Pipeline
+func PipelineNamespace(namespace string) PipelineOp {
+	return func(p *v1alpha1.Pipeline) {
+		p.ObjectMeta.Namespace = namespace
+	}
+}
+
+// PipelineSpec sets the PipelineSpec to the Pipeline.
+// Any number of PipelineSpec modifier can be passed to transform it.
+func PipelineSpec(ops ...PipelineSpecOp) PipelineOp {
+	return func(p *v1alpha1.Pipeline) {
+		ps := &p.Spec
+
+		for _, op := range ops {
+			op(ps)
+		}
+
+		p.Spec = *ps
+	}
+}
+
+// PipelineTask adds a PipelineTask, with specified name and task name, to the PipelineSpec.
+// Any number of PipelineTask modifier can be passed to transform it.
+func PipelineTask(name, taskName string, ops ...PipelineTaskOp) PipelineSpecOp {
+	return func(ps *v1alpha1.PipelineSpec) {
+		pTask := &v1alpha1.PipelineTask{
+			Name: name,
+			TaskRef: v1alpha1.TaskRef{
+				Name: taskName,
+			},
+		}
+		for _, op := range ops {
+			op(pTask)
+		}
+		ps.Tasks = append(ps.Tasks, *pTask)
+	}
+}
+
+// PipelineRun creates a PipelineRun with default values.
+// Any number of PipelineRun modifier can be passed to transform it.
+func PipelineRun(name string, ops ...PipelineRunOp) *v1alpha1.PipelineRun {
+	pr := &v1alpha1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: defaultNamespace,
+			Name:      name,
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineTriggerRef: v1alpha1.PipelineTriggerRef{
+				Type: v1alpha1.PipelineTriggerTypeManual,
+			},
+		},
+	}
+
+	for _, op := range ops {
+		op(pr)
+	}
+
+	return pr
+}
+
+// PipelineRunNamespace sets the namespace of the PipelineRun.
+func PipelineRunNamespace(namespace string) PipelineRunOp {
+	return func(p *v1alpha1.PipelineRun) {
+		p.ObjectMeta.Namespace = namespace
+	}
+}
+
+// PipelineRunSpec sets the PipelineRunSpec, with specifed name, to the PipelineRun.
+// Any number of PipelineRunSpec modifier can be passed to transform it.
+func PipelineRunSpec(name string, ops ...PipelineRunSpecOp) PipelineRunOp {
+	return func(pr *v1alpha1.PipelineRun) {
+		prs := &pr.Spec
+
+		prs.PipelineRef.Name = name
+
+		for _, op := range ops {
+			op(prs)
+		}
+
+		pr.Spec = *prs
+	}
+}
+
+// PipelineRunServiceAccount sets the service account to the PipelineRunSpec.
+func PipelineRunServiceAccount(sa string) PipelineRunSpecOp {
+	return func(prs *v1alpha1.PipelineRunSpec) {
+		prs.ServiceAccount = sa
+	}
+}
+
+// PipelineResource creates a PipelineResource with default values.
+// Any number of PipelineResource modifier can be passed to transform it.
+func PipelineResource(name string, ops ...PipelineResourceOp) *v1alpha1.PipelineResource {
```

Comment:
How about updating `PipelineResource ` definition to

```
func PipelineResource(name, namespace string, ops ... PipelineResourceOp) *v1alpha1. PipelineResource {
```

Like above comment, lets not assume namespace for objects